### PR TITLE
Adds Gregory surfaces to `primal`

### PR DESCRIPTION
**Pull request by jcs15c**
# Summary

This PR adds `GregoryPatch` and `GregoryTriangle` classes to primal, which are surface objects which can be defined only through a set of boundary curves and tangent vectors at the interior control points of those curves. The classes are meant to mirror `BezierPatch` and `BezierTriangle` respectively. These changes will be followed by methods that do more interesting things to Gregory patches, such as ray-casting and/or winding numbers methods.

This has required minor changes to other geometric objects:
 - A degree elevation method for Bezier curves

### CODE REVIEW

File: RELEASE-NOTES.md [hunk 1]
```diff
@@ -57,6 +57,7 @@ The Axom project release numbers follow [Semantic Versioning](http://semver.org/
 - Klee: Adds support for lua-based input decks for shaping
 - Slam: Adds convenience aliases in `axom/slam/Aliases.hpp` for the most common set and relation configurations, 
   including `ArraySet`, `ArrayViewSet`, `VariableRelation`, `ConstantRelation` and their `View` forms.
+- Primal: Adds Gregory surface classes, with a `primal::GregoryPatch` class for cubic quadrilateral surfaces and a `primal::GregoryTriangle` class for hybrid cubic-quartic triangle surfaces.
```
Comment (kennyweiss) [inline RELEASE-NOTES.md]: Please don't forget to move this to the `Unreleased` section above the `0.15 release` section after updating to axom@develop

File: src/axom/primal/geometry/BezierCurve.hpp [hunk 2]
```diff
@@ -395,6 +395,91 @@ class BezierCurve
...
+    {
+      return;
+    }
+
+    int ord = getOrder();
+    if(ord < 0)
+    {
+      return;
+    }
+
```
Comment (kennyweiss) [inline src/axom/primal/geometry/BezierCurve.hpp]: Optimization opportunity (not necessarily required for this branch though): 
we can reserve the updated control points array space (and weights) for `getOrder()+degrees`

We could also reserve the fully sized temp arrays at the outset.

File: src/axom/primal/tests/primal_bezier_curve.cpp [hunk 3]
```diff
@@ -232,6 +232,50 @@ TEST(primal_beziercurve, evaluate)
   }
 }
 
+//------------------------------------------------------------------------------
+TEST(primal_beziercurve, degree_elevate_preserves_geometry)
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_bezier_curve.cpp]: Nice test!

File: src/axom/primal/tests/primal_gregory_surfaces.cpp [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+
+  for(int i = 0; i < 4; ++i)
+  {
+    EXPECT_EQ(patch.getCorner(i), corners[i]);
+  }
+
+  EXPECT_EQ(patch.evaluate(0.0, 0.0), corners[0]);
+  EXPECT_EQ(patch.evaluate(1.0, 0.0), corners[1]);
+  EXPECT_EQ(patch.evaluate(1.0, 1.0), corners[2]);
+  EXPECT_EQ(patch.evaluate(0.0, 1.0), corners[3]);
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_gregory_surfaces.cpp]: Should this also test points within edges and in the interior of the patch?

File: src/axom/primal/tests/primal_gregory_surfaces.cpp
Comment (kennyweiss) [inline src/axom/primal/tests/primal_gregory_surfaces.cpp]: It seems like we need more tests that check interior points of Gregory patches/triangles generated from edge data + normals.

Perhaps this can wait for a follow-up PR though.

File: src/axom/primal/geometry/GregoryPatch.hpp [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+   * \param [in] bPatch A polynomial Bezier patch of order (3, 3)
+   *
+   * This creates a Gregory patch that exactly reproduces the input bicubic Bezier patch.
+   * The Gregory tangent pairs are duplicated from the four Bezier interior control points,
+   * causing the parameter-dependent Gregory blends to collapse to fixed Bezier points.
+   *
+   * \pre \a bPatch must have order (3, 3)
+   * \pre \a bPatch must be polynomial, not rational
+   */
+  explicit GregoryPatch(const BezierPatch<T, 3>& bPatch)
```
Comment (kennyweiss) [inline src/axom/primal/geometry/GregoryPatch.hpp]: Not sure if this would be useful/necessary, but if the orders of `bPatch` were less than 3 in `u` or `v`, we could degree-elevat `bPatch`.

File: src/axom/primal/geometry/GregoryPatch.hpp [hunk 6]
```diff
@@ -0,0 +1,758 @@
...
+
+      getTangent(i, 0) = PointType(x1.array() + b1.array());
+      getTangent(i, 1) = PointType(x2.array() + b2.array());
+    }
+  }
+
+  ///@}
+
+  /// \brief Returns the \a i-th corner point, oriented ccw
+  PointType& getCorner(int i) { return m_controlPoints[i]; }
```
Comment (kennyweiss) [inline src/axom/primal/geometry/GregoryPatch.hpp]: Please update the docs to indicate that $0<=i<4$.

Should we add `SLIC_ASSERT_IF` debug checks on the indexing, here and in `getTangent()` ?

Comment (kennyweiss) [reply src/axom/primal/geometry/GregoryPatch.hpp]: Update -- there are a bunch of functions below that should indicate the valid ranges in the doxygen and add debug checks

File: src/axom/primal/geometry/GregoryPatch.hpp [hunk 7]
```diff
@@ -0,0 +1,758 @@
...
+  {
+    os << "GregoryPatch(";
+    for(int i = 0; i < NPTS; ++i)
+    {
+      os << m_controlPoints[i];
+      if(i + 1 < NPTS)
+      {
+        os << ", ";
+      }
+    }
```
Comment (kennyweiss) [inline src/axom/primal/geometry/GregoryPatch.hpp]: Should this split the cases by corner/edge/tangent?
